docs/embedded/tutorials/migrate-abs-to-spe.md

```diff
@@ -0,0 +1,320 @@
+    await foreach (var blobItem in _containerClient.GetBlobsAsync())
+    {
+        blobs.Add(blobItem.Name);
+    }
+    return blobs;
+```
+
+### Thread pooling
+```C#
+    private CountdownEvent _countdown;
+
+    // This is how the thread pool knows how many files are being migrated
+    _countdown = new CountdownEvent(blobs.Count);
+```
+
+### Traverse blob list
+```C#
+    // It creates a new folder in the destination. The name of the folder is the blob's container name.
+    containerFolder = await _gcm.CreateFolder(_containerName, "root");
+
+    // Traverse the blob list
+    foreach (var blobName in fileList)
+    {
+        FileStructure fs = new FileStructure() { blobName = blobName };
+
+        // This function parses the flat file into folder hierarchy and create the folder structure in the destination. It will retrieve the parentFolderId that the file should be copied to.
+        // If you are going to copy it to root you can comment this line out. The parentFolderId will be containerFolder.Id
+        fs.parentFolderId = TraverseFileListing(fs, containerFolder.Id)
+
+        // This is where the thread pool happens
+        ThreadPool.QueueUserWorkItem(MigrateFile, fs);
+    }
+
+    // Call so the program doesn't end, it waits for all the files to be processed
+    _countdown.Wait();
+```
+
+### Traverse file listing
+```C#
+    // Check if file exist before traversing to same time
+    ...
+
+    // Parse for folder path not including the file name and put it in an array
+    var pathSegments = filePath.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
+    string[] directoriesParts = pathSegments.Take(pathSegments.Length - 1).ToArray();
+
+    // Traverse the folder listing and create 1 folder at a time
+    string relativePath = _containerName;
+    string newFolderId = parentFolderId;
+    foreach (string folderName in directoriesParts)
+    {
+        string newPath = relativePath + _separator + folderName;
+        ...
+
+        DriveItem subFolder = await _gcm.CheckIfItemExists(folderName, newFolderId);
+        if (subFolder == null)
+        {
+            subFolder = await _gcm.CreateFolder(folderName, newFolderId);
+            ...
+        }
+        newFolderId = subFolder.Id;
+
+        relativePath = newPath;
+    }
+
+    return newFolderId;
+```
+
+### Check if item exists
+```C#
+    var item = await _graphClient.Drives[_containerId].Items[parentFolderId].ItemWithPath(itemPath).GetAsync();
+```
+
+### Create folder
+```C#
+    var folder = new DriveItem
+    {
+        Name = folderName,
+        Folder = new Folder(),
+        AdditionalData = new Dictionary<string, object>()
+        {
+            { "@microsoft.graph.conflictBehavior", "fail" }
+        }
+    };
+    var createdFolder = await _graphClient.Drives[_containerId].Items[parentFolderId].Children.PostAsync(folder);
+```
+
+### MigrateFile
+```C#
+    internal async void MigrateFile(Object stateInfo)
```

Comment:
Why is this being passed in as an Object type and then converted to a FilesStructure type? Can an interface be implement or it just be a File Structure type?